core/domain/exp_jobs_one_off.py

```diff
@@ -77,6 +78,89 @@
 AUDIO_FILE_PREFIX = 'audio'
 AUDIO_ENTITY_TYPE = 'exploration'
 AUDIO_DURATION_SECS_MIN_STATE_SCHEMA_VERSION = 31
+# This threshold puts a cap on the number of valid inputs, i.e.,
+# expressions/equations that can be yielded by the math expression one-off jobs.
+# The reason for limiting the number of valid inputs yielded by the jobs is that
+# we don't need to closesly inspect all of the valid inputs, they are displayed
+# just to make sure that the job output is what we expect.
+VALID_MATH_INPUTS_YIELD_LIMIT = 200
+_TYPE_INVALID_EXPRESSION = 'Invalid'
+_TYPE_VALID_ALGEBRAIC_EXPRESSION = 'AlgebraicExpressionInput'
+_TYPE_VALID_NUMERIC_EXPRESSION = 'NumericExpressionInput'
+_TYPE_VALID_MATH_EQUATION = 'MathEquationInput'
+
+
+def _clean_math_expression(math_expression):
+    """Cleans a given math expression and formats it so that it is compatible
+    with the new interactions' validators.
+
+    Args:
+        math_expression: str. The string representing the math expression.
+
+    Returns:
+        str. The correctly formatted string representing the math expression.
+    """
+    unicode_to_text = {
+        u'\u221a': 'sqrt',
+        u'\xb7': '*',
+        u'\u03b1': 'alpha',
+        u'\u03b2': 'beta',
+        u'\u03b3': 'gamma',
+        u'\u03b4': 'delta',
+        u'\u03b5': 'epsilon',
+        u'\u03b6': 'zeta',
+        u'\u03b7': 'eta',
+        u'\u03b8': 'theta',
+        u'\u03b9': 'iota',
+        u'\u03ba': 'kappa',
+        u'\u03bb': 'lambda',
+        u'\u03bc': 'mu',
+        u'\u03bd': 'nu',
+        u'\u03be': 'xi',
+        u'\u03c0': 'pi',
+        u'\u03c1': 'rho',
+        u'\u03c3': 'sigma',
+        u'\u03c4': 'tau',
+        u'\u03c5': 'upsilon',
+        u'\u03c6': 'phi',
+        u'\u03c7': 'chi',
+        u'\u03c8': 'psi',
+        u'\u03c9': 'omega',
+    }
+    inverse_trig_fns_mapping = {
+        'asin': 'arcsin',
+        'acos': 'arccos',
+        'atan': 'arctan'
+    }
+    trig_fns = ['sin', 'cos', 'tan', 'csc', 'sec', 'cot']
+
+    # Shifting powers in trig functions to the end.
+    # For eg. 'sin^2(x)' -> '(sin(x))^2'.
+    for trig_fn in trig_fns:
+        math_expression = re.sub(
+            r'%s(\^\d)\((.)\)' % trig_fn,
+            r'(%s(\2))\1' % trig_fn, math_expression)
+
+    # Adding parens to trig functions that don't have
+    # any. For eg. 'cosA' -> 'cos(A)'.
+    for trig_fn in trig_fns:
+        math_expression = re.sub(
+            r'%s(?!\()(.)' % trig_fn, r'%s(\1)' % trig_fn, math_expression)
+
+    # The pylatexenc lib outputs the unicode values of
+    # special characters like sqrt and pi, which is why
+    # they need to be replaced with their corresponding
+    # text values before performing validation.
```

Comment:
Add: Other unicode characters will be left in the string as-is, and will be rejected by the expression parser.

Reply:
Done.